core/src/main/java/org/elasticsearch/discovery/zen/NodeJoinController.java

```diff
@@ -376,20 +378,26 @@ public ClusterState execute(ClusterState currentState) {
                 while (iterator.hasNext()) {
                     Map.Entry<DiscoveryNode, List<MembershipAction.JoinCallback>> entry = iterator.next();
                     final DiscoveryNode node = entry.getKey();
-                    joinCallbacksToRespondTo.addAll(entry.getValue());
                     iterator.remove();
-                    if (currentState.nodes().nodeExists(node.getId())) {
+                    final String preflight = nodesBuilder.preflightPut(node);
+                    if (preflight != null) {
+                        Throwable failure = new IllegalStateException(preflight);
+                        for (MembershipAction.JoinCallback callback: entry.getValue()) {
+                            joinCallbacksToFail.add(new Tuple<>(callback, failure));
+                        }
+                    } else if (currentState.nodes().nodeExists(node.getId())) {
```

Comment:
If a node with same persisted node id but different address tries to join the cluster (because the data folder was copied from another node) it will not manage to join the cluster. It will not have any meaningful error message either except that it couldn't join. Can we tell the offending node why it failed using joinCallbacksToFail?